### PR TITLE
[WIP] Array bounds checks

### LINKED ISSUE #110: Array out of bounds access
**IR codegen** will need a decision on this topic.

### PR DESCRIPTION
**Pull request by joshuabach**
Closes #110 

Implement bound checks for arrays. It can be controlled with the `"array_bounds_checks"` feature flag (which is enabled by the `"safe"` feature (which is enabled by default).

I should probably add some tests, but I wanted to wait for #166

### CODE REVIEW

**Comment by reiner-dolp** [comment]
a few integration tests for this would be cool.

File: compiler-lib/src/firm/method_body_generator.rs [hunk 1]
```diff
@@ -302,7 +307,11 @@ impl<'a, 'ir, 'src, 'ast> MethodBodyGenerator<'ir, 'src, 'ast> {
             Some(mode) => Value(ValueComputation::simple(
                 &call.project_result_tuple().project(mode, 0),
             )),
-            None => Void,
+            None => {
+                // TODO is this problematic?
```
Comment (problame) [inline compiler-lib/src/firm/method_body_generator.rs]: don't know if `this` referrs to the keep_alive, but I guess the keep_alive is problematic at some point

File: compiler-lib/src/firm/method_body_generator.rs [hunk 2]
```diff
@@ -542,36 +551,63 @@ impl<'a, 'ir, 'src, 'ast> MethodBodyGenerator<'ir, 'src, 'ast> {
...
+                let checked_array_type = &self.type_analysis.expr_info(expr).ty;
+                let array_type = ty_from_checked_type(checked_array_type)
+                    .expect("array type must have firm equivalent")
+                    .points_to();
+                let elt_type = ty_from_checked_type(
+                    checked_array_type
+                        .inner_type()
+                        .expect("type of array must have inner type"),
+                )
+                .expect("array element type must have firm equivalent");
```
Comment (problame) [inline compiler-lib/src/firm/method_body_generator.rs]: indentation weird

File: compiler-lib/src/firm/method_body_generator.rs [hunk 3]
```diff
@@ -542,36 +551,63 @@ impl<'a, 'ir, 'src, 'ast> MethodBodyGenerator<'ir, 'src, 'ast> {
...
                     self.graph.cur_store(),
                     self.graph.new_addr(self.runtime.new),
                     &[alloc_size.as_value_node()],
                 );
                 self.graph.set_store(call.project_mem());
 
-                Value(ValueComputation::simple(
-                    &call.project_result_tuple().project(unsafe { mode::P }, 0),
-                ))
+                // This is the pointer to our array structure (consisting of len+data)
```
Comment (problame) [inline compiler-lib/src/firm/method_body_generator.rs]: conditionalize the comment?

File: compiler-lib/src/firm/method_body_generator.rs [hunk 4]
```diff
@@ -595,12 +632,61 @@ impl<'a, 'ir, 'src, 'ast> MethodBodyGenerator<'ir, 'src, 'ast> {
...
+            self.graph.cur_block().mature();
+            self.graph.set_cur_block(err_block);
+
+            let res = self.gen_static_fn_call(
+                self.runtime.array_out_of_bounds,
+                None,
+                &[idx_expr, len.as_value_node()],
+            );
+            assert_matches!(res, ExprResult::Void);
+            unsafe { keep_alive(err_block.into()) };
```
Comment (problame) [inline compiler-lib/src/firm/method_body_generator.rs]: remove before lowering

**Comment by problame** [comment]
> @problame Can you verify that [4cc718f](https://github.com/comprakt/comprakt/commit/4cc718f5e684369fb159870fd63f12367c895ebb) is correct?

the comment makes sense to me, although I don't have your state-of-mind, in particular which test cases lead to that change...

**Comment by joshuabach** [comment]
This is blocked until the code gen is refactored to @hediet's new firm wrappers

**Comment by hediet** [comment]
I can give it another try, as the libfirm bug allegedly has been fixed.

**Comment by problame** [comment]
Can this one be closed in favor of #220 ?

**Comment by joshuabach** [comment]
> Can this one be closed in favor of #220 ?

Yes.

/close

**Pull request closed without merging**